clash-cores/src/Clash/Cores/CRC/Internal.hs

```diff
@@ -0,0 +1,650 @@
+
+      hwParams = CRCHardwareParams
+        { _crcNlanes = nLanes
+        , _crcBaseParams = crcParams crc dataWidth
+        , _crcLaneParams = mkCRCLaneParams crc dataWidth nLanes
+        , _crcResidues = fmap computeResidue $ reverse indicesI
+        }
+
+-- | Matrix multiplication in the binary finite field
+matVecMul
+  :: forall (m :: Nat)
+            (n :: Nat)
+   . KnownNat m
+  => KnownNat n
+  -- ^ @n@ columns
+  => Vec m (BitVector n)
+  -- ^ The Matrix
+  -> BitVector n
+  -- ^ The column vector
+  -> BitVector m
+  -- ^ The resulting vector
+matVecMul mat vec = v2bv $ fmap (\row -> reduceXor $ vec .&. row) mat
+
+-- | Like `subSNat` but uses a `Data.Type.Ord.<=` constraint
+subSNatLe
+  :: forall (n :: Nat)
+            (m :: Nat)
+   . m <= n
+  => SNat n
+  -> SNat m
+  -> SNat (n - m)
+subSNatLe a b = leToPlus @m @n $ subSNat a b
+
+-- | Like `take` but uses a `Data.Type.Ord.<=` constraint
+takeLe
+  :: forall (n :: Nat)
+            (m :: Nat)
+            (a :: Type)
+   . n <= m
+  => SNat n
+  -- ^ How many elements to take
+  -> Vec m a
+  -- ^ input vector
+  -> Vec n a
+takeLe SNat vs = leToPlus @n @m $ takeI vs
+
+-- | Apply a step to a CRC and input given by @lanePrev@
+--   This function crashes if called with out of bounds @lanePrev@
+--   meant to be used via `smap`
+laneStep
+  :: CRCHardwareParams crcWidth dataWidth nLanes
+  -> SNat lanePrev
+  -> ()
+  -> BitVector crcWidth
+  -> Vec nLanes (BitVector dataWidth)
+  -> BitVector crcWidth
+laneStep (CRCHardwareParams nLanes@SNat (CRCParams SNat SNat _ _ _ _ _) laneParams _) lanePrev@SNat _ crc input
+  = let lane = addSNat lanePrev d1 in case compareSNat lane nLanes of
+    SNatLE -> matVecMul (getFGMatrix laneParams lane) (crc ++# (pack $ takeLe lane input))
+    _      -> clashCompileError "lookupLaneStep: Absurd, Report this to the Clash compiler team: https://github.com/clash-lang/clash-compiler/issues"
+
+-- | A parallel multilane CRC engine and validator
+--
+-- @dataWidth@ _should_ always cleanly divide @crcWidth@. If this is not the case
+-- then it is possible for the last message to be misaligned.
+--
+-- For example if you have @dataWidth@ of @48@ and a @crcWidth@ of @32@. Then a possible
+-- message could be 128-bits, two words + CRC. But this cannot be safely
+-- ran through the CRC circuit. Because the final fragment contains 32 valid bits
+-- and 16 invalid bits.
+--
+-- This rule is __NOT__ checked at the type level because as long you take
+-- care that misalignment does not occur it's not a problem.
+crcEngine
+  :: forall (dom :: Domain)
+            (crc :: Type)
+            (dataWidth :: Nat)
+            (nLanes :: Nat)
+   . HiddenClockResetEnable dom
+  => HardwareCRC crc dataWidth nLanes
+  => Proxy crc
+  -- ^ The CRC
+  -> Proxy dataWidth
+  -- ^ The width of the words to feed
+  -> Proxy nLanes
+  -- ^ How many words can be fed at maximum in a single cycle
```

Comment:
Erm, I was too quick in scanning tyvars. The `Proxy crc` is still needed, because `crc` is not constrained, it is only _used_ in the form of `CRCWidth crc`. But I think my point stands for `nLanes` and `dataWidth`.